lib/IntegrationEvent/Implementation/AWS-SNS-SQS/RedisHelper.ts

```diff
@@ -0,0 +1,76 @@
+import { IntegrationEvent } from "../../IntegrationEvent";
+import { WrappedNodeRedisClient } from "handy-redis";
+
+export class RedisHelper {
+  public static async storeEventData(options: {
+    event: IntegrationEvent;
+    RedisClient: WrappedNodeRedisClient;
+  }): Promise<void> {
+    if (options.event && options.event.queueId) {
+      try {
+        const hashKey = `{${options.event.queueId}}_data`;
+        const sortedSetKey = `{${options.event.queueId}}_queue`;
+        const masterSetKey = `master_event_queue`;
+        const eventId = options.event.integrationEventId;
+        const eventData = JSON.stringify(options.event);
+        const score = new Date().getTime();
+        await options.RedisClient.multi()
+          .zadd(sortedSetKey, [score, eventId])
+          .hsetnx(hashKey, eventId, eventData)
+          .exec();
+        await options.RedisClient.zadd(masterSetKey, "NX", [score, sortedSetKey]);
+      } catch (error) {
+        console.log(`Error while saving event to Redis: ${options.event.integrationEventId}`);
+      }
+    }
+  }
+
+  public static async canStartEventExecution(options: {
+    event: IntegrationEvent;
+    RedisClient: WrappedNodeRedisClient;
+  }): Promise<boolean> {
+    if (options.RedisClient && options.event && options.event.queueId) {
+      const activeSetKey = `{${options.event.queueId}}_active`;
+      const activeEventKey = `{${options.event.queueId}}_${options.event.integrationEventId}_active`;
+      const keysExist = await options.RedisClient.exists(activeEventKey, activeSetKey);
+      if (keysExist > 0) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public static async fetchNextEvent(options: {
+    queueId: string;
+    RedisClient: WrappedNodeRedisClient;
+  }): Promise<string | undefined> {
+    if (options.RedisClient && options.queueId) {
+      const activeSetKey = `{${options.queueId}}_active`;
+      const hashKey = `{${options.queueId}}_data`;
+      const sortedSetKey = `{${options.queueId}}_queue`;
+      const eventIdList = await options.RedisClient.zpopmin(sortedSetKey, 1);
+      let activeEventKey;
+      if (eventIdList.length > 0) {
+        const eventId = eventIdList[0];
+        const eventDataString = await options.RedisClient.hget(hashKey, eventId);
+        activeEventKey = `{${options.queueId}}_${eventId}_active`;
+        if (eventDataString) {
+          await options.RedisClient.multi()
+            .setnx(activeSetKey, "1")
+            .setnx(activeEventKey, "1")
+            .expire(activeEventKey, 120)
+            .exec();
+          return eventDataString;
+        }
+      } else {
+        await options.RedisClient.multi()
+          .del(sortedSetKey, hashKey, activeSetKey)
+          .exec();
+        const queueIdList = await options.RedisClient.zpopmin(`master_event_queue`, 1);
```

Comment:
@al-polarr Mani and I had a discussion about this. Basically, before events are published, the corresponding queue IDs are stored in Redis's master queue set. So there are two ways to deal with how to get the next queue to process when you finish processing all events of the current queue:

1. Mani's current method of handling this is to pop the earliest queue ID from the master queue, then process all events under that queue ID.
2. We can use SQS to initiate the processing of a queue. Considering that SQS is at least once delivery, the event itself can be seen as a mutex, once delivered, it should trigger the processing of a corresponding event queue.

I was thinking method 1 renders SQS effectively useless except in the cases where:

1. It is the first event arriving on a consumer, triggering the `fetchNextEvent` infinite loop.
2. Redis crashes, and we would rely on SQS's loosely FIFO event delivery, where if an event arrives and does not exist in Redis, the consumers will process it anyway.

My concern being method 1 might waste compute cycles by keep checking `fetchNextEvent`, and the master queue set in Redis becomes central orchestration point. If we scale up, we will end up with many consumers trying to `zpopmin` on the master queue set. Another point being technically SQS should be more reliable than Redis, since in Redis we are constrained by memory/sharding etc.

Instead I thought we should let events coming from SQS to initiate queue processing, when any event from a queue arrives, start processing all events in that queue, from the earliest to the oldest, and when the queue has emptied, exit the loop and wait for the next initiation. However, if there are any consistency issues, we may get events in the master queue that's never emptied. We may have to build a master queue police bot to check on master queue consistency, and if any events are not processed within a certain timeframe(say 10 minutes), process them and alert us.

We'd like to hear your thoughts on this, and are still brainstorming possible edge cases with either solutions.

Reply:
The need for coupling Redis and SQS is not immediately obvious from looking at this implementation.  It would help to have concrete examples of interaction patterns that give rise to these multiple queue IDs to provide understanding of how the events must subsequently be de-queued from multiple queue IDs to ensure correctness.  With that understanding in place, it becomes more straightforward to determine if the implementation is correct, and write automated tests to validate that correctness.  To your point about edge cases, enumerating interaction patterns in this way will also help us identify any gaps that might not be captured here.

Reply:
@al-polarr Provided 2 examples in https://www.notion.so/polarr/Redis-FIFO-Queue-RFC-a4160df6fb9346cea9c8ee95396c5efd Let me know if this helps.

Reply:
I've also scheduled a meeting for tomorrow so we can sync up about this.